vocode/streaming/action/nylas_send_email.py

```diff
@@ -1,22 +1,39 @@
-import json
-from vocode.streaming.action.base_action import BaseAction
+from typing import Optional, Type
+from pydantic import Field
 import os
-from vocode.streaming.models.actions import NylasSendEmailActionOutput
+from vocode.streaming.action.respond_action import RespondAction
+from vocode.streaming.models.actions import ActionInput, ActionOutput, ActionType
 
 
-class NylasSendEmail(BaseAction[NylasSendEmailActionOutput]):
-    def run(self, params: str) -> NylasSendEmailActionOutput:
-        """Sends an email using Nylas API.
-        The input to this action is a pipe separated list of the recipient email, email body, optional subject. But always include
-        the pipe character even if the subject or message IDs are not included and just leave it blank.
-        The subject should only be included if it is a new email thread.
-        If there is no message id, the email will be sent as a new email. Otherwise, it will be sent as a reply to the given message. Make sure to include the previous message_id
-        if you are replying to an email.
+class NylasSendEmailActionInput(ActionInput):
+    class Parameters(ActionInput.Parameters):
+        recipient_email: str = Field(
+            ..., description="The email address of the recipient."
+        )
+        body: str = Field(..., description="The body of the email.")
+        subject: Optional[str] = Field(None, description="The subject of the email.")
+
+    action_type: str = ActionType.NYLAS_SEND_EMAIL.value
+    params: Parameters
+
+
+class NylasSendEmailActionOutput(ActionOutput):
+    class Response(ActionOutput.Response):
+        success: bool
+
+    action_type: str = ActionType.NYLAS_SEND_EMAIL.value
+    response: Response
 
-        For example, `recipient@example.com|Hello, this is the email body.|this is the subject` would send an email to recipient@example.com with the provided body and subject.
-        """
-        recipient_email, email_body, email_subject = params.split("|")
 
+class NylasSendEmail(
+    RespondAction[NylasSendEmailActionInput, NylasSendEmailActionOutput]
+):
+    description: str = "Sends an email using Nylas API."
+    action_type: str = ActionType.NYLAS_SEND_EMAIL.value
```

Comment:
You may want to consider NOT having an in-library registry of action types: `ActionType`.  I'm thinking about what process someone would follow if they wanted to use vocode with their own special action.  They _can_ define their own class `class DoMySpecialThing`, which derives from `RespondAction` and has its own corresponding input/output types, and they can provide this class when configuring their ActionAgent.  But the one thing they can't do easily is revise `ActionType`, since that's a part of the library itself.

Reply:
`action_type` is a string, so you can just make an action like:

```
class MyAction:
  action_type = "my_action"
```

in terms of the `ActionFactory`, similar to `AgentFactory`, you would define your own `ActionFactory`

```
class MyActionFactory:
  
  def create_action(self, action_type: str):
    if action_type == "my_action": return MyAction()
...
```

then you would pass the `ActionFactory` into the `ActionAgent`

```
agent = ActionAgent(prompt_preamble=..., action_factory=MyActionFactory())
```

Reply:
adding to this.. without the `ActionFactory` the agent would have no way to create the action/call the action